### DISCUSSION TITLE
Plagin to get a collection from database

### DISCUSSION THREAD

**Discussion by agolovan** [payloadcms/payload#1841] [Feature Requests & Ideas]
We are planning to store collections into database and allow non coders from separate UI to modify this collection. Each of this collections actually a document. This is how it looks:
![modelcollections](https://user-images.githubusercontent.com/19178208/211494120-980f1191-c1ae-4dc0-b5cb-cc1fe626dba1.JPG)

I want to create a plugin that allow to get collection from database and merge with collections that are defined in typescript, This is how plugin looks with hardcoded external collection:
const myCollections: Array<CollectionConfig> = [
  {
    slug: 'MilaCollection',
    admin: {
      useAsTitle: 'Eddie',
    },
    access: {},
    fields: [
      {
        name: 'Eddie',
        type: 'text',
        localized: true,
        maxLength: 99999,
        minLength: 1,
        index: true,
        unique: true,
      },
      {
        name: 'Larisa',
        type: 'text',
        localized: true,
        maxLength: 99999,
        minLength: 1,
        index: true,
        unique: true,
      },
    ],
  },
];

const addCollectionsFromDatabase: Plugin = (incomingConfig: Config): Config => {
  // Spread the existing config
  const config: Config = {
    ...incomingConfig,
    collections: [
      ...incomingConfig.collections, ...myCollections
    ]  
  };

  return config;
};

export default addCollectionsFromDatabase;

However, in place of myCollections, I need to use async/await and get data from database. I am getting CollectionConfig[] type from database, how I could use that in this plugin?

**Comment by jacobsfletch** [comment]
@agolovan this makes perfect sense to me. You are needing to await a response from your external database before building the Payload config. Sounds like we need to build the plugin configs asynchronously. Here's an entry-point into this: https://github.com/payloadcms/payload/blob/96002dbda5e28565e9bbc3927dfd985798859e36/src/config/build.ts

This sounds entirely possible but I'll do some digging here.

**Reply by jmikrut** [reply]
With the change to `payload.init` becoming async in the next release, this will be 100% possible.

Before, `async` plugins were impossible, but very shortly this will become doable.

**Comment by agolovan** [comment]
How could I get a response from external database when using plagin as it is running at web browser? At this time it is under a web application that could write/read to MonboDB, could we use same API?

**Reply by jacobsfletch** [reply]
You need to alias your server-only code so that it does not run in the browser, here are the docs related to that: https://payloadcms.com/docs/admin/webpack#aliasing-server-only-modules

You can also see a good example of this in the Stripe plugin here: https://github.com/payloadcms/plugin-stripe/blob/main/src/extendWebpackConfig.ts
